Fix certificate issue with Spoke and Encrypted Hub

WARN 2015-03-09 17:59:40,054 [pool-6-thread-1] com.flightstats.hub.spoke.RemoteSpokeStore [line 111] - unable to get content aav2_pfTdmItineraries/2015/03/09/17/58/59/991/a2sLK8
com.sun.jersey.api.client.ClientHandlerException: javax.net.ssl.SSLHandshakeException: java.security.cert.CertificateException: No subject alternative DNS name matching encrypted-hub-v2-03.cloud-east.staging found.
	at com.sun.jersey.client.urlconnection.URLConnectionClientHandler.handle(URLConnectionClientHandler.java:155) ~[jersey-client-1.19.jar:1.19]
	at com.sun.jersey.api.client.Client.handle(Client.java:652) ~[jersey-client-1.19.jar:1.19]
	at com.sun.jersey.api.client.WebResource.handle(WebResource.java:682) ~[jersey-client-1.19.jar:1.19]
	at com.sun.jersey.api.client.WebResource.get(WebResource.java:193) ~[jersey-client-1.19.jar:1.19]
	at com.flightstats.hub.spoke.RemoteSpokeStore.read(RemoteSpokeStore.java:102) ~[hub-2015-02-17.29.jar:na]
	at com.flightstats.hub.spoke.SpokeContentDao.read(SpokeContentDao.java:68) [hub-2015-02-17.29.jar:na]
	at com.flightstats.hub.dao.s3.S3WriteQueue.writeContent(S3WriteQueue.java:75) [hub-2015-02-17.29.jar:na]
	at com.flightstats.hub.dao.s3.S3WriteQueue.access$100(S3WriteQueue.java:23) [hub-2015-02-17.29.jar:na]
	at com.flightstats.hub.dao.s3.S3WriteQueue$2.call(S3WriteQueue.java:62) [hub-2015-02-17.29.jar:na]
	at com.flightstats.hub.dao.s3.S3WriteQueue$2.call(S3WriteQueue.java:59) [hub-2015-02-17.29.jar:na]
	at com.github.rholder.retry.AttemptTimeLimiters$NoAttemptTimeLimit.call(AttemptTimeLimiters.java:78) [guava-retrying-1.0.6.jar:na]
	at com.github.rholder.retry.Retryer.call(Retryer.java:110) [guava-retrying-1.0.6.jar:na]
	at com.flightstats.hub.dao.s3.S3WriteQueue.write(S3WriteQueue.java:59) [hub-2015-02-17.29.jar:na]
	at com.flightstats.hub.dao.s3.S3WriteQueue.access$000(S3WriteQueue.java:23) [hub-2015-02-17.29.jar:na]
	at com.flightstats.hub.dao.s3.S3WriteQueue$1.call(S3WriteQueue.java:50) [hub-2015-02-17.29.jar:na]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_31]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_31]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_31]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_31]
Caused by: javax.net.ssl.SSLHandshakeException: java.security.cert.CertificateException: No subject alternative DNS name matching encrypted-hub-v2-03.cloud-east.staging found.
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:192) ~[na:1.8.0_31]
	at sun.security.ssl.SSLSocketImpl.fatal(SSLSocketImpl.java:1937) ~[na:1.8.0_31]
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:302) ~[na:1.8.0_31]
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:296) ~[na:1.8.0_31]
	at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1478) ~[na:1.8.0_31]
	at sun.security.ssl.ClientHandshaker.processMessage(ClientHandshaker.java:212) ~[na:1.8.0_31]
	at sun.security.ssl.Handshaker.processLoop(Handshaker.java:957) ~[na:1.8.0_31]
	at sun.security.ssl.Handshaker.process_record(Handshaker.java:892) ~[na:1.8.0_31]
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:1050) ~[na:1.8.0_31]
	at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1363) ~[na:1.8.0_31]
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1391) ~[na:1.8.0_31]
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1375) ~[na:1.8.0_31]
	at sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:563) ~[na:1.8.0_31]
	at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:185) ~[na:1.8.0_31]
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1512) ~[na:1.8.0_31]
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1440) ~[na:1.8.0_31]
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480) ~[na:1.8.0_31]
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:338) ~[na:1.8.0_31]
	at com.sun.jersey.client.urlconnection.URLConnectionClientHandler._invoke(URLConnectionClientHandler.java:253) ~[jersey-client-1.19.jar:1.19]
	at com.sun.jersey.client.urlconnection.URLConnectionClientHandler.handle(URLConnectionClientHandler.java:153) ~[jersey-client-1.19.jar:1.19]
	... 18 common frames omitted
Caused by: java.security.cert.CertificateException: No subject alternative DNS name matching encrypted-hub-v2-03.cloud-east.staging found.

--- a/src/main/java/com/flightstats/hub/group/GroupCaller.java
+++ b/src/main/java/com/flightstats/hub/group/GroupCaller.java
@@ -8,6 +8,7 @@ import com.flightstats.hub.cluster.LastContentKey;
 import com.flightstats.hub.cluster.Leader;
 import com.flightstats.hub.metrics.MetricsTimer;
 import com.flightstats.hub.model.ContentKey;
+import com.flightstats.hub.rest.RestClient;
 import com.flightstats.hub.util.RuntimeInterruptedException;
 import com.github.rholder.retry.RetryException;
 import com.github.rholder.retry.Retryer;
@@ -89,7 +90,7 @@ public class GroupCaller implements Leader {
             logger.info("group is missing, exiting " + group.getName());
             return;
         }
-        this.client = GroupClient.createClient();
+        this.client = RestClient.createClient(30, 120);
         callbackQueue = queueProvider.get();
         try {
             ContentKey lastCompletedKey = lastContentKey.get(group.getName(), new ContentKey(), GROUP_LAST_COMPLETED);

--- a/src/main/java/com/flightstats/hub/rest/RestClient.java
+++ b/src/main/java/com/flightstats/hub/rest/RestClient.java
@@ -1,4 +1,4 @@
-package com.flightstats.hub.group;
+package com.flightstats.hub.rest;
 
 import com.sun.jersey.api.client.Client;
 import com.sun.jersey.api.client.config.ClientConfig;
@@ -17,11 +17,11 @@ import java.security.cert.X509Certificate;
 import java.util.concurrent.TimeUnit;
 
 
-class GroupClient {
+public class RestClient {
 
-    private final static Logger logger = LoggerFactory.getLogger(GroupClient.class);
+    private final static Logger logger = LoggerFactory.getLogger(RestClient.class);
 
-    public static Client createClient() {
+    public static Client createClient(int connectTimeout, int readTimeout) {
         try {
             TrustManager[] certs = new TrustManager[]{
                     new X509TrustManager() {
@@ -49,8 +49,8 @@ class GroupClient {
             config.getProperties().put(HTTPSProperties.PROPERTY_HTTPS_PROPERTIES,
                     new HTTPSProperties((hostname, session) -> true, ctx));
             Client client = Client.create(config);
-            client.setConnectTimeout((int) TimeUnit.SECONDS.toMillis(30));
-            client.setReadTimeout((int) TimeUnit.SECONDS.toMillis(120));
+            client.setConnectTimeout((int) TimeUnit.SECONDS.toMillis(connectTimeout));
+            client.setReadTimeout((int) TimeUnit.SECONDS.toMillis(readTimeout));
             client.setFollowRedirects(true);
             return client;
         } catch (Exception e) {

--- a/src/main/java/com/flightstats/hub/spoke/RemoteSpokeStore.java
+++ b/src/main/java/com/flightstats/hub/spoke/RemoteSpokeStore.java
@@ -7,6 +7,7 @@ import com.flightstats.hub.model.Content;
 import com.flightstats.hub.model.ContentKey;
 import com.flightstats.hub.model.Trace;
 import com.flightstats.hub.model.Traces;
+import com.flightstats.hub.rest.RestClient;
 import com.google.common.base.Optional;
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import com.google.inject.Inject;
@@ -32,7 +33,7 @@ public class RemoteSpokeStore {
 
     private final static Logger logger = LoggerFactory.getLogger(RemoteSpokeStore.class);
 
-    private final static Client client = create();
+    private final static Client client = RestClient.createClient(5, 5);
 
     private final SpokeCluster cluster;
     private final MetricsSender sender;
@@ -45,13 +46,6 @@ public class RemoteSpokeStore {
         executorService = Executors.newCachedThreadPool(new ThreadFactoryBuilder().setNameFormat("RemoteSpokeStore-%d").build());
     }
 
-    private static Client create() {
-        Client client = Client.create();
-        client.setConnectTimeout((int) TimeUnit.SECONDS.toMillis(5));
-        client.setReadTimeout((int) TimeUnit.SECONDS.toMillis(5));
-        return client;
-    }
-
     public boolean write(String path, byte[] payload, Content content) throws InterruptedException {
         List<String> servers = cluster.getServers();
         int quorum = getQuorum(servers.size());
@@ -61,10 +55,10 @@ public class RemoteSpokeStore {
             executorService.submit(new Runnable() {
                 @Override
                 public void run() {
-                    content.getTraces().add(new Trace(server));
+                    String uri = HubHost.getScheme() + server + "/spoke/payload/" + path;
+                    content.getTraces().add(new Trace(uri));
                     try {
-                        ClientResponse response = client.resource(HubHost.getScheme() + server + "/spoke/payload/" + path)
-                                .put(ClientResponse.class, payload);
+                        ClientResponse response = client.resource(uri).put(ClientResponse.class, payload);
                         long complete = System.currentTimeMillis();
                         content.getTraces().add(new Trace(server, response.getEntity(String.class)));
                         if (response.getStatus() == 201) {

